typegen/src/utils.rs

```diff
@@ -131,6 +131,21 @@ fn types_equal_inner(
     let a_generic_idx = a_parent_params.index_for_type_id(a);
     let b_generic_idx = b_parent_params.index_for_type_id(b);
 
+    let a_ty = types.resolve(a).expect("type a should exist in registry");
+    let b_ty = types.resolve(b).expect("type b should exist in registry");
+
+    // Capture a few variables to avoid some repetition later when we recurse.
+    let mut types_equal_recurse =
+        |a: u32, a_params: &GenericsList, b: u32, b_params: &GenericsList| -> bool {
+            types_equal_inner(a, a_params, a_visited, b, b_params, b_visited, types)
+        };
+
+    // We'll lazily extend our type params only if the shapes match.
+    let calc_params = || {
+        let a_params = a_parent_params.extend(&a_ty.type_params);
+        let b_params = b_parent_params.extend(&b_ty.type_params);
+        (a_params, b_params)
+    };
     // If both IDs map to same generic param, then we'll assume equal. If they don't
```

Comment:
Nit; add back a newline above :)

```suggestion

    // If both IDs map to same generic param, then we'll assume equal. If they don't
```